### PR TITLE
[v1.x] Implemented MKLDNN Backward Adaptive Pooling kernel

### PR DESCRIPTION
**Pull request by piotrwolinski-intel**
## Description ##
This PR introduces MKLDNN backward adaptive pooling implementation introduced to master in [this](https://github.com/apache/incubator-mxnet/pull/20825) PR.

### CODE REVIEW

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [website]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [website]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [website, unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [website, unix-cpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Review by bgawrych — APPROVED** [review]

File: src/operator/nn/pooling-inl.h [hunk 1]
```diff
@@ -166,6 +196,10 @@ struct hash<mxnet::op::PoolingParam> {
     ret = dmlc::HashCombine(ret, val.count_include_pad);
     int val_layout = val.layout.has_value() ? val.layout.value() : -1;
     ret = dmlc::HashCombine(ret, val_layout);
+    mxnet::Tuple<int> val_out_size = val.output_size.has_value()
```
Comment (RafLit) [inline src/operator/nn/pooling-inl.h:199]: shouldn't IsAdaptivePooling be used here?

File: src/operator/contrib/adaptive_avg_pooling.cc [hunk 2]
```diff
@@ -27,11 +27,13 @@
 #include "../operator_common.h"
 #include "adaptive_avg_pooling-inl.h"
 #if MXNET_USE_MKLDNN == 1
+#include "../nn/mkldnn/mkldnn_base-inl.h"
 #include "../nn/mkldnn/mkldnn_pooling-inl.h"
 #endif  // MXNET_USE_MKLDNN
 
 #define START_IND(a, b, c) static_cast<int>(std::floor(static_cast<float>(a * c) / b))
 #define END_IND(a, b, c) static_cast<int>(std::ceil(static_cast<float>((a + 1) * c) / b))
+#define DIV_ROUND_UP(a, b) ((a + (b - 1)) / b)
```
[the remark this answers is not in the document]
Comment (piotrwolinski-intel) [reply src/operator/contrib/adaptive_avg_pooling.cc]: Done

File: src/operator/contrib/adaptive_avg_pooling.cc [hunk 3]
```diff
@@ -219,10 +221,10 @@ bool SupportMKLDNNAveragePooling(const NDArray &in_data,
...
-  const int strides_H = floor((IH << 1) / OH) - floor(IH / OH);
-  const int strides_W = floor((IW << 1) / OW) - floor(IW / OW);
-  const int kernel_H = ceil((IH << 1) / OH) - floor(IH / OH);
-  const int kernel_W = ceil((IW << 1) / OW) - floor(IW / OW);
+  const int strides_H = ((IH << 1) / OH) - (IH / OH);
+  const int strides_W = ((IW << 1) / OW) - (IW / OW);
+  const int kernel_H = DIV_ROUND_UP((IH << 1) / OH, 1) - (IH / OH);
+  const int kernel_W = DIV_ROUND_UP((IW << 1) / OW, 1) - (IW / OW);
   const int pad_l_top = (strides_H * (OH - 1) + kernel_H - IH) / 2;
   const int pad_l_left = (strides_W * (OW - 1) + kernel_W - IW) / 2;
```
[the remark this answers is not in the document]
Comment (piotrwolinski-intel) [reply src/operator/contrib/adaptive_avg_pooling.cc]: Separated them into functions

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-gpu]

**Review by RafLit — APPROVED** [review]

**Pull request merged**